core/src/main/scala/org/apache/spark/util/collection/PartitionedSerializedPairBuffer.scala

```diff
@@ -246,9 +247,13 @@ private[spark] class SerializedSortDataFormat extends SortDataFormat[Int, IntBuf
 }
 
 private[spark] object PartitionedSerializedPairBuffer {
-  val KEY_START = 0
-  val VAL_START = 1
-  val VAL_END = 2
+  val KEY_START = 0 // keyStart, a long, gets split across two ints
+  val KEY_VAL_LEN = 2
   val PARTITION = 3
-  val RECORD_SIZE = Seq(KEY_START, VAL_START, VAL_END, PARTITION).size // num ints of metadata
+  val RECORD_SIZE = PARTITION + 1 // num ints of metadata
+
+  def keyStartPos(metaBuffer: IntBuffer, metaBufferPos: Int): Long = {
+    metaBuffer.get(metaBufferPos + KEY_START) +
```

Comment:
The lower 32 bits of the long might be interpreted as a negative integer here, so I think that this might give an incorrect result when added to the shifted upper 32 bits.